auth/auth0_manager.go

```diff
@@ -1,23 +1,95 @@
 package auth
 
 import (
-	"log"
+	"fmt"
+	"time"
 
+	"github.com/zpatrick/go-cache"
 	"github.com/zpatrick/rclient"
 )
 
 type Auth0Manager struct {
-	token  string
-	client *rclient.RestClient
+	clientID   string
+	connection string
+	client     *rclient.RestClient
+	cache      *cache.Cache
 }
 
-func NewAuth0Manager(endpoint, token string) *Auth0Manager {
+type Auth0Config struct {
+	Domain     string
+	ClientID   string
+	Connection string
+}
+
+type oauthReq struct {
+	ClientID   string `json:"client_id"`
+	Connection string `json:"connection"`
+	Username   string `json:"username"`
+	Password   string `json:"password"`
+	GrantType  string `json:"grant_type"`
+	Scope      string `json:"scope"`
+}
+
+type authStatus struct {
+	isValid bool
+	penalty time.Duration
+}
+
+const (
+	maxPenalty      = 5 * time.Second
+	validAuthExpiry = 1 * time.Hour
+)
+
+func NewAuth0Manager(config Auth0Config) *Auth0Manager {
 	return &Auth0Manager{
-		client: rclient.NewRestClient(endpoint),
+		clientID:   config.ClientID,
+		connection: config.Connection,
+		client:     rclient.NewRestClient(config.Domain),
+		cache:      cache.New(),
 	}
 }
 
-func (a *Auth0Manager) Authenticate(user, pass string) (bool, error) {
-	log.Println("[ERROR] - Auth0Manager.Authenticate not implemented")
+func (a *Auth0Manager) Authenticate(username, password string) (bool, error) {
+	key := fmt.Sprintf("%s:%s", username, password)
+	var cachedStatus *authStatus
+	if result, exists := a.cache.Getf(key); exists {
+		cachedStatus = result.(*authStatus)
+	} else {
+		cachedStatus = &authStatus{}
+	}
+
+	if cachedStatus.isValid {
+		return true, nil
+	}
+
+	// will only sleep if cachedStatus already exists with a penalty
+	time.Sleep(cachedStatus.penalty)
```

Comment:
Ok. So let me try to talk through this so I understand it properly… 

When a service attempts to pull from `d.ims.io` and fails, the container will fail and a new one is spawned soon after, thus creating a new http client connection to `d.ims.io`. As such, I’m not clear on how sleeping a single client connection would solve our DDoS issue?